x11rb/src/rust_connection/stream.rs

```diff
@@ -456,76 +447,27 @@ impl Stream for DefaultStream {
 
             let fd = self.as_raw_fd();
             let msg = loop {
-                #[cfg(any(
-                    target_os = "android",
-                    target_os = "dragonfly",
-                    target_os = "freebsd",
-                    target_os = "linux",
-                    target_os = "netbsd",
-                    target_os = "openbsd"
-                ))]
-                let flags = MsgFlags::MSG_CMSG_CLOEXEC;
-                #[cfg(not(any(
-                    target_os = "android",
-                    target_os = "dragonfly",
-                    target_os = "freebsd",
-                    target_os = "linux",
-                    target_os = "netbsd",
-                    target_os = "openbsd"
-                )))]
-                let flags = MsgFlags::empty();
-
-                match recvmsg::<()>(fd, &mut iov, Some(&mut cmsg), flags) {
+                match recvmsg::<()>(fd, &mut iov, Some(&mut cmsg), recvmsg::flags()) {
                     Ok(msg) => break msg,
                     // try again
                     Err(nix::Error::EINTR) => {}
                     Err(e) => return Err(e.into()),
                 }
             };
 
-            let mut cloexec_error = None;
             let fds_received = msg
                 .cmsgs()
                 .flat_map(|cmsg| match cmsg {
                     ControlMessageOwned::ScmRights(r) => r,
                     _ => Vec::new(),
                 })
-                .filter_map(|fd| {
-                    // set cloexec if we didn't before
-                    #[cfg(not(any(
-                        target_os = "android",
-                        target_os = "dragonfly",
-                        target_os = "freebsd",
-                        target_os = "linux",
-                        target_os = "netbsd",
-                        target_os = "openbsd"
-                    )))]
-                    let res = fcntl::fcntl(fd, fcntl::FcntlArg::F_SETFD(fcntl::FdFlag::FD_CLOEXEC))
-                        .map(|_| fd);
-                    #[cfg(any(
-                        target_os = "android",
-                        target_os = "dragonfly",
-                        target_os = "freebsd",
-                        target_os = "linux",
-                        target_os = "netbsd",
-                        target_os = "openbsd"
-                    ))]
-                    let res = nix::Result::Ok(fd);
-
-                    match res {
-                        Ok(fd) => Some(RawFdContainer::new(fd)),
-                        Err(e) => {
-                            cloexec_error = Some(e);
-                            None
-                        }
-                    }
-                });
-            fd_storage.extend(fds_received);
+                .map(RawFdContainer::new);
 
-            match cloexec_error {
-                None => Ok(msg.bytes),
-                Some(e) => Err(e.into()),
-            }
+            let mut cloexec_error = Ok(());
```

Comment:
If you want, the helper function could switch to `flat_map` and just start dropping things when `cloexec_error` is set. However, I doubt it really makes much of a difference in practice. Does anyone ever receive more than one FD at the same time? Does anyone even use FD passing? :see_no_evil:

Reply:
Yeah, FD passing is a cold operation in my opinion. I think the `extend` implementation for `Vec` even unrolls the iteration by one to avoid allocation, so in most cases an empty iterator is the optimal case.

Reply:
I didn't even know that X11 had FD passing until I found x11rb